File: ts/indicators/welles-wilder/average-true-range/average-true-range.ts
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorType } from '../../core/indicator-type';
import { OutputMetadata } from '../../core/outputs/output-metadata';
import { OutputType } from '../../core/outputs/output-type';
import { TrueRange } from '../true-range/true-range';
import { AverageTrueRangeOutput } from './average-true-range-output';

const atrMnemonic = 'atr';
const atrDescription = 'Average True Range';

/**
 * Welles Wilder's Average True Range (ATR) indicator.
 *
 * ATR averages True Range (TR) values over the specified length using the Wilder method:
 * - multiply the previous value by (length - 1)
 * - add the current TR value
 * - divide by length
 *
 * The initial ATR value is a simple average of the first length TR values.
 * The indicator is not primed during the first length updates.
 */
export class AverageTrueRange implements Indicator {

  private readonly length_: number;
  private readonly lastIndex: number;
  private stage = 0;
  private windowCount = 0;
  private window: number[];
  private windowSum = 0;
  private value_ = NaN;
  private primed_ = false;
  private readonly trueRange: TrueRange;

  constructor(length: number) {
    if (length < 1) {
      throw new Error(`invalid length ${length}: must be >= 1`);
    }

    this.length_ = length;
    this.lastIndex = length - 1;
    this.window = this.lastIndex > 0 ? new Array<number>(length) : [];
    this.trueRange = new TrueRange();
  }

  /** Returns the length parameter. */
  public get length(): number {
    return this.length_;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean {
    return this.primed_;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    const outputMeta: OutputMetadata = {
      kind: AverageTrueRangeOutput.AverageTrueRangeValue,
      type: OutputType.Scalar,
      mnemonic: atrMnemonic,
      description: atrDescription,
    };

    return {
      type: IndicatorType.AverageTrueRange,
      mnemonic: atrMnemonic,
      description: atrDescription,
      outputs: [outputMeta],
    };
  }

  /** Updates the Average True Range given the next bar's close, high, and low values. */
  public update(close: number, high: number, low: number): number {
    if (isNaN(close) || isNaN(high) || isNaN(low)) {
      return NaN;
    }

    const trueRangeValue = this.trueRange.update(close, high, low);

    if (this.lastIndex === 0) {
      this.value_ = trueRangeValue;

      if (this.stage === 0) {
        this.stage++;
      } else if (this.stage === 1) {
        this.stage++;
        this.primed_ = true;
      }

      return this.value_;
    }

    if (this.stage > 1) {
      // Wilder smoothing method.
      this.value_ *= this.lastIndex;
      this.value_ += trueRangeValue;
      this.value_ /= this.length_;
      return this.value_;
    }

    if (this.stage === 1) {
      this.windowSum += trueRangeValue;
      this.window[this.windowCount] = trueRangeValue;
      this.windowCount++;

      if (this.windowCount === this.length_) {
        this.stage++;
        this.primed_ = true;
        this.value_ = this.windowSum / this.length_;
      }

      return this.primed_ ? this.value_ : NaN;
    }

    // The very first sample is used by the True Range.
    this.stage++;
    return NaN;
  }

  /** Updates the Average True Range using a single sample value as a substitute for high, low, and close. */
  public updateSample(sample: number): number {
    return this.update(sample, sample, sample);
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const v = sample.value;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(sample.close, sample.high, sample.low);
    return [scalar];
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = (sample.bid + sample.ask) / 2;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const v = sample.price;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }
}

File: go/indicators/welleswilder/averagetruerange/averagetruerange.go
package averagetruerange

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/welleswilder/truerange"
)

const (
	atrMnemonic    = "atr"
	atrDescription = "Average True Range"
)

// AverageTrueRange is Welles Wilder's Average True Range indicator.
//
// ATR averages True Range (TR) values over the specified length using the Wilder method:
//   - multiply the previous value by (length - 1)
//   - add the current TR value
//   - divide by length
//
// The initial ATR value is a simple average of the first length TR values.
// The indicator is not primed during the first length updates.
//
// Reference:
//
// Wilder, J. Welles Jr. (1978). New Concepts in Technical Trading Systems.
type AverageTrueRange struct {
	mu          sync.RWMutex
	length      int
	lastIndex   int
	stage       int
	windowCount int
	window      []float64
	windowSum   float64
	value       float64
	primed      bool
	trueRange   *truerange.TrueRange
}

// NewAverageTrueRange returns a new instance of the Average True Range indicator.
func NewAverageTrueRange(length int) (*AverageTrueRange, error) {
	if length < 1 {
		return nil, fmt.Errorf("invalid length %d: must be >= 1", length)
	}

	atr := &AverageTrueRange{
		length:    length,
		lastIndex: length - 1,
		value:     math.NaN(),
		trueRange: truerange.NewTrueRange(),
	}

	if atr.lastIndex > 0 {
		atr.window = make([]float64, length)
	}

	return atr, nil
}

// Length returns the length parameter.
func (s *AverageTrueRange) Length() int {
	return s.length
}

// IsPrimed indicates whether the indicator is primed.
func (s *AverageTrueRange) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *AverageTrueRange) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.AverageTrueRange,
		atrMnemonic,
		atrDescription,
		[]core.OutputText{
			{Mnemonic: atrMnemonic, Description: atrDescription},
		},
	)
}

// Update updates the Average True Range given the next bar's close, high, and low values.
func (s *AverageTrueRange) Update(close, high, low float64) float64 {
	if math.IsNaN(close) || math.IsNaN(high) || math.IsNaN(low) {
		return math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	trueRangeValue := s.trueRange.Update(close, high, low)

	if s.lastIndex == 0 {
		s.value = trueRangeValue

		if s.stage == 0 {
			s.stage++
		} else if s.stage == 1 {
			s.stage++
			s.primed = true
		}

		return s.value
	}

	if s.stage > 1 {
		// Wilder smoothing method.
		s.value *= float64(s.lastIndex)
		s.value += trueRangeValue
		s.value /= float64(s.length)

		return s.value
	}

	if s.stage == 1 {
		s.windowSum += trueRangeValue
		s.window[s.windowCount] = trueRangeValue
		s.windowCount++

		if s.windowCount == s.length {
			s.stage++
			s.primed = true
			s.value = s.windowSum / float64(s.length)
		}

		if s.primed {
			return s.value
		}

		return math.NaN()
	}

	// The very first sample is used by the True Range.
	s.stage++

	return math.NaN()
}

// UpdateSample updates the Average True Range using a single sample value
// as a substitute for high, low, and close.
func (s *AverageTrueRange) UpdateSample(sample float64) float64 {
	return s.Update(sample, sample, sample)
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *AverageTrueRange) UpdateScalar(sample *entities.Scalar) core.Output {
	v := sample.Value

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (s *AverageTrueRange) UpdateBar(sample *entities.Bar) core.Output {
	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(sample.Close, sample.High, sample.Low)}

	return output
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *AverageTrueRange) UpdateQuote(sample *entities.Quote) core.Output {
	v := (sample.Bid + sample.Ask) / 2 //nolint:mnd

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *AverageTrueRange) UpdateTrade(sample *entities.Trade) core.Output {
	v := sample.Price

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}
